Flip it if you can't collapse it

sloriot opened the pull request:
if a needle is also a cap, it will be handled as a needle only. But if we can't collapse the edge, a flip will not be tried. Try the flip too.

Note that this will slow down the method as the status of a triangle is tested twice.

Fix #8605

sloriot committed:
in case an edge is both a cap and a needle but that is not collapsable try to flip

sloriot committed:
add missing cap collection in case of non-collapsed edge

sloriot committed:
split badly shaped triangle check and delay the check calls

sloriot committed:
swap the parameters also here

MaelRL commented on Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:
```diff
@@ -112,14 +151,17 @@ void collect_badly_shaped_triangles(const typename boost::graph_traits<TriangleM
+    if (res[1] == boost::graph_traits<TriangleMesh>::null_halfedge())
+      std::cout << "add new needle: " << edge(res[0], tmesh) << std::endl;
+    else
+      std::cout << "add new needle (also a cap): " << edge(res[0], tmesh) << std::endl;
```
Do you not fill both ranges on purpose? There's still:

```
  ...
  else // let's not make it possible to have a face be both a cap and a needle (for now)
 ...
```

a little below; so, if it's a needle & a cap, you fill the needle range, but not the cap range.

MaelRL committed:
Fix collect_badly_shaped_triangles not actually filling both ranges

MaelRL committed:
Fix never trying again a user-rejected cap

MaelRL committed:
Avoid a lot of useless shape predicate calls

Here are some changes:
- Avoid a lot of redundant calls of is_it_a_needle / is_it_a_cap
  - Do not test for cap-ness if it's already a needle
  - Do not fill both ranges at the beginning, but fill
    with everything: on the first round, needle-ness will
    be tested, and the exit towards cap-ness will happen
  - Wait for a needle to no longer be a needle, or not be
    treatable to actually test cap-ness
- When the mesh is modified, do not test everything immediately,
  just put them in the queue of the next iteration

Result: Fewer calls to shape predicates

- master -
needle calls = 828 668
cap calls = 803 330

- PR -
needle calls = 803 554
cap calls = 795 317

Should be fewer but two effects are balacing each other for the calls:
- much fewer calls to shape predicates (is_it_a_needle, is_it_a_cap)
  by not calling until absolutely necessary (but still checking at pop time).
- when we modify the mesh, I no longer fill the CURRENT cap and needle
  ranges, but the next ones. The point is to really prioritize all caps
  BEFORE treating needles, whereas otherwise we would treat some needles
  at the current iteration while there are still caps to treat.
So there are more iterations which add more useless calls (see below)

MaelRL committed:
Misc cleaning

MaelRL committed:
Fix test pretty much never testing envelope tests + read all formats

MaelRL commented:
Here are some changes:
- Avoid a lot of redundant calls of is_it_a_needle / is_it_a_cap
  - Always avoid testing for cap-ness if it's already a needle
  - Wait for a needle to no longer be a needle, or not be treatable to actually test cap-ness
  - No need to fill both ranges at the beginning: if we start from all faces, needle-ness will be tested, and the exit towards cap-ness will happen if its not a needle
- When the mesh is modified, do not test everything immediately, but put them in the queue of the next iteration (which will test for needle-ness through the "is it still a needle?" mechanism)

Result: Fewer calls to shape predicates; for example

- master -
needle calls = 828 668
cap calls = 803 330

- PR -
needle calls = 803 554
cap calls = 795 317

Should be fewer but the base there is a fixed base cost at 735k faces and two effects are balancing each other for the calls:
- much fewer calls to shape predicates (is_it_a_needle, is_it_a_cap) by not calling until absolutely necessary (but still checking at pop time).
- when we modify the mesh, I no longer fill the CURRENT cap and needle ranges, but the next ones. The point is to really prioritize all caps   BEFORE treating needles, whereas otherwise we would treat some needles at the current iteration while there are still caps to treat.

MaelRL commented:
Gives a slight improvement on the quality too (PR = before last batch of commits, TMP = after last batch of commits):

![Screenshot From 2024-12-07 15-11-28](https://github.com/user-attachments/assets/630462a0-3924-42b7-9b2b-621de7dfa9da)

sloriot committed:
only use one container for the next loop

--> make the output of the function idempotent

sloriot committed:
factorize cap checks

sloriot committed:
do not put all the faces of the mesh in the hash map

sloriot commented:
Successfully tested in CGAL-6.1-Ic-45
